entity-service-client/src/main/java/org/hypertrace/entity/data/service/client/EdsCacheClient.java

```diff
@@ -45,12 +54,31 @@ public EdsCacheClient(EntityDataServiceClient client) {
             });
 
     this.entityCache =
-        CacheBuilder.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).maximumSize(1000)
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(cacheConfig.getEntityCacheExpiryMs(), TimeUnit.MILLISECONDS)
+            .maximumSize(cacheConfig.getEntityMaxCacheSize())
             .build(new CacheLoader<>() {
               public Optional<Entity> load(@Nonnull EdsCacheKey key) {
                 return Optional.ofNullable(client.getById(key.tenantId, key.entityId));
               }
             });
+
+    this.entityIdsCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(cacheConfig.getEntityIdsCacheExpiryMs(), TimeUnit.MILLISECONDS)
+        .maximumSize(cacheConfig.getEntityIdsMaxCacheSize())
+        .build(new CacheLoader<>() {
+          public Optional<String> load(@Nonnull EdsTypeAndIdAttributesCacheKey key) {
+            Entity entity = client.getByTypeAndIdentifyingAttributes(key.tenantId,
+                key.byTypeAndIdentifyingAttributes);
+            String entityId = null;
+            if (entity != null) {
+              entityId = entity.getEntityId();
+              entityCache.put(new EdsCacheKey(entity.getTenantId(), entity.getEntityId()),
```

Comment:
So, you're updating `entityCache` in addition to this cache when there is a hit?

Reply:
It will be handle as any other entry. The property `expireAfterWriteNanos` is the cache level and its apply to all entries.

Reply:
Why do you need to explicitly add .put for LoadingCache? This could cause race-condition. It is better to leverage the cache's mechanism and call .invalidate. The next time a consumer calls "get" it will load from DB.

Reply:
We already have entity cache which is based on EdsCacheKey. We need one more based on identifying attributes. So, to avoid duplicates, we have added additional cache to store only entityIds (not full Entity) which are loaded via function - `getByTypeAndIdentifyingAttributes`. And, to avoid additional functional call back to DB which was done `getById` inside `getByTypeAndIdentifyingAttributes`, we are updating entity cache. LocalCache is already synchronized using a lock. So, I think, this should be fine.

Reply:
We are putting into a different cache. Not sure what you mean by invalidate helping

Reply:
I see. Yes, in this case invalidate wouldn't help and as for the race-condition, you're not computing before putting so that wouldn't lead to the problem.